test/cpp/end2end/xds/xds_override_host_end2end_test.cc

```diff
@@ -0,0 +1,241 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+#include "src/core/ext/filters/client_channel/backup_poller.h"
+#include "src/core/lib/gprpp/match.h"
+#include "src/proto/grpc/testing/xds/v3/cluster.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/outlier_detection.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/router.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/stateful_session.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/stateful_session_cookie.pb.h"
+#include "test/core/util/scoped_env_var.h"
+#include "test/cpp/end2end/xds/xds_end2end_test_lib.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+using ::envoy::extensions::filters::http::stateful_session::v3::StatefulSession;
+using ::envoy::extensions::filters::network::http_connection_manager::v3::
+    HttpFilter;
+using ::envoy::extensions::http::stateful_session::cookie::v3 ::
+    CookieBasedSessionState;
+
+class OverrideHostTest : public XdsEnd2endTest {
+ protected:
+  struct Cookie {
+    std::string value;
+    std::set<std::string> attributes;
+    std::string raw;
+  };
+
+  static std::map<std::string, Cookie, std::less<>> ParseCookies(
+      const std::map<std::string, std::string, std::less<>>& initial_metadata) {
+    auto it = initial_metadata.find("set-cookie");
+    if (it == initial_metadata.end()) {
+      return {};
+    }
+    // Right now we only support a single cookie. This will be revised as
+    // semantics are better understood
+    std::pair<absl::string_view, absl::string_view> name_value =
+        absl::StrSplit(it->second, absl::MaxSplits('=', 1));
+    if (name_value.first.empty()) {
+      return {};
+    }
+    std::pair<absl::string_view, absl::string_view> value_attrs =
+        absl::StrSplit(name_value.second, absl::MaxSplits(';', 1));
+    std::set<std::string> attributes;
+    for (absl::string_view segment : absl::StrSplit(name_value.second, ';')) {
+      attributes.insert(std::string(absl::StripAsciiWhitespace(segment)));
+    }
+    return {{std::string(name_value.first),
+             Cookie({std::string(value_attrs.first), std::move(attributes),
+                     std::string(name_value.second)})}};
+  }
+
+  static absl::optional<std::string> GetStatefulSessionCookie(
+      const std::map<std::string, std::string, std::less<>>& initial_metadata,
+      absl::string_view cookie_name = "grpc_session_cookie") {
```

Comment:
Suggest defining a global constant at the top of the file for the cookie name, so that we don't need to repeat it all over the place:

```
constexpr absl::string_view kCookieName = "grpc_session_cookie";
```